### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] False positive where removing the type assertion leads to TS error

### ISSUE THREAD

**Issue by marcelgerber** [typescript-eslint/typescript-eslint#7207]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.1.6&fileType=.tsx&code=DYUwLgBMCWCioFsQDswGcBcEASAVAsgDKECS8ISqA2gLoBQdokIiaEAvBAIIBOPAhgE8AdADMeAewQAKACYSAxgFdKYYQEclIHoIDKLEArASeXYMGkAiflGiWAlPYj82eIqXKrajOIhToOCAMENCA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
let liElements: HTMLLIElement[]

let elems = Array.from(document.querySelectorAll("a li")) as HTMLLIElement[]
liElements = elems
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expected no error to appear here, since removing the `as HTMLLIElement` leads to a TS error.

### Actual Result

TSEslint suggests removing the `as HTMLLIElement` (including auto-fixing it), which then leads to a TypeScript error in line 4:

```
Type 'Element[]' is not assignable to type 'HTMLLIElement[]'.
  Type 'Element' is missing the following properties from type 'HTMLLIElement': type, value, accessKey, accessKeyLabel, and 119 more.(2322)
let liElements: HTMLLIElement[]
```

### Additional Info

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `6.0.0` |
| `@typescript-eslint/parser` | `6.0.0` |
| `TypeScript` | `5.1.6` |
| `ESLint` | `8.39.0` |
| `node` | `web` |

**Comment by bradzacher** [comment]
This is the classic problem of a generic that's only used in a return location.

You have `Array.from<T>(arg: Iterable<T>): T` - the returned type comes from the input.

However your input is `querySelectorAll<T>(): T`...

So combining the two the generic returned from `Array.from` comes from the inner call.

TS behaves very strangely when there is a generic that's only used in a return location - the TLDR is that if you assert the expression it backfills the generic with the asserted type.

I.e. It does this `querySelectorAll<Foo>() as Foo`, rather than using the default for the generic and then changing it with the assertion.

This leads to a problem because when we interrogate the types it means we see `T as T` - which is obviously an unnecessary condition!

So TLDR again - this is working as intended due to limitations with TS. You can fix this by moving the assertion to be a generic annotation:

```ts
Array.from(document.querySelectorAll<HTMLLIElement>("a li"))
```
